Document client.escapeIdentifier and client.escapeLiteral

TheConner opened the pull request:
Per #1978 it seems that these client APIs are undocumented. I added documentation for these methods along with some examples and relevant links.

I wasn't too sure what kind of structure to take with the documentation, such as documenting the usage via a client instance and the Client prototype. I can shuffle things around if there is a better way of expressing the usage.

charmander commented on docs/pages/apis/client.mdx:
````diff
@@ -256,6 +257,44 @@ client
+
+First, this can be used on an existing instance of `Client`:
+```js
+const escpaedIdentifier = client.escapeIdentifier('FooIdentifier')
+console.log(escpaedIdentifier) // '"FooIdentifier"'
````
```suggestion
const escapedIdentifier = client.escapeIdentifier('FooIdentifier')
console.log(escapedIdentifier) // '"FooIdentifier"'
```

charmander commented on docs/pages/apis/client.mdx:
````diff
@@ -256,6 +257,44 @@ client
+First, this can be used on an existing instance of `Client`:
+```js
+const escapedLiteral = client.escapeLiteral("hello 'world'")
+console.log(escapedLiteral) // "Hello ''world''"
````
```suggestion
console.log(escapedLiteral) // "'hello ''world'''"
```

charmander commented on docs/pages/apis/client.mdx:
````diff
@@ -256,6 +257,44 @@ client
+Alternatively, this can be used via `Client.prototype`:
+```js
+const { Client } = require('pg')
+const escpaedIdentifier = Client.prototype.escapeIdentifier('Bar"Identifier')
+console.log(escpaedIdentifier) // '"Bar""Identifier"'
````
```suggestion
const escapedIdentifier = Client.prototype.escapeIdentifier('Bar"Identifier')
console.log(escapedIdentifier) // '"Bar""Identifier"'
```

charmander reviewed (changes requested):
I think we should export these as `pg.escapeIdentifier` and `pg.escapeLiteral` instead of documenting the `Client.prototype` workaround.

TheConner commented:
Fixed the typos, thank you for pointing them out! 

> I think we should export these as `pg.escapeIdentifier` and `pg.escapeLiteral` instead of documenting the `Client.prototype` workaround.

Agreed. It should be fairly simple to do this, and it would be cleaner from a documentation perspective. I'll follow up with those changes

brianc commented on packages/pg/test/unit/client/escape-tests.js:
```diff
@@ -1,65 +0,0 @@
-'use strict'
```
deleting tests means breaking, non-backwards compatible changes.  We shouldn't do this.  For legacy reasons you should leave the methods on `client` but only document them on the main export `pg` object. They need to be properly deprecated and removed in a multi-month long release cycle.

TheConner replied:
Works for me. For what it is worth the tests were moved from client to util, but it is still a breaking change. I wasn't sure if implementing charmander's suggestions would be applicable to this PR, since it was a breaking API change. I'll make these changes backwards compatible, thus no changes to unit tests will be needed.

brianc reviewed (changes requested):
I like the thinking here, but the best thing to do is leave them on both `client` and also export them on the root `pg` object and only document them on the `pg.` object. Removing stuff & causing backwards incompatible changes is a months long process w/ deprecation warnings and coordinating other breaking changes and weighing their impact. For the sake of the millions of users the best thing to do is keep it on both `Client.prototype` and add it to `pg` and then in another PR we could deprecate the old usage.

brianc commented:
awesome! thank you - yeah we can just have the functions available in both places (with tests for both access paths) & we're good to goooo! I appreciate the work here - over the years I just got very, very cautious w/ breaking changes. :)

TheConner commented:
@brianc no worries! a future PR or issue can be created for the breaking changes

Updated with requested changes -- these updates are now backwards compatible. This is a matter of preference, but I chose to keep the functions in utils but re-export on client with a disclaimer that it is for backwards compatibility. Restored original tests for client escapes, added an additional test step for the other entry point in utils. Not sure how I feel about testing the second entry point in the client test suite, but the alternative is duplicating the tests for the utils test suite.

charmander commented on packages/pg/lib/client.js:
```diff
@@ -456,36 +456,12 @@ class Client extends EventEmitter {
+  // escapeIdentifier and escapeLiteral moved to utility functions & exported
+  // on PG 
+  // re-exported here for backwards compatibility
+  escapeIdentifier = utils.escapeIdentifier
```
Need to be defined on the prototype, not as a class field, to avoid breaking backwards compatibility with people who were already using `Client.prototype` directly. Sorry, I should have mentioned that initially.

TheConner replied:
Right, that was one of the workarounds people are using. I'll fix that, and also add test cases that cover the usage of Client.prototype.escape*, as currently they only cover usages of instantiated clients

charmander commented on packages/pg/lib/index.js:
```diff
@@ -23,6 +24,8 @@ var PG = function (clientConstructor) {
   this.Connection = Connection
   this.types = require('pg-types')
   this.DatabaseError = DatabaseError
+  this.escapeIdentifier = escapeIdentifier
+  this.escapeLiteral = escapeLiteral
```
This is just new though, right? Why?

TheConner replied:
This is new yes, as for why [your previous comment](https://github.com/brianc/node-postgres/pull/2954#pullrequestreview-1378415905) suggested that the escape* methods be exported on pg, this adds them as exports to the pg object. On my end it looks like the github diff viewer annoyingly cuts off the part where it shows these belonging to PG, but looking at the [full file](https://github.com/brianc/node-postgres/blob/fe01e3db7f578039a9f4a1daa4cc3d5816c4cb2e/packages/pg/lib/index.js#L18-L29) it shows the rest of the context.

Is there an issue with exporting them this way?
